pyproject.toml

```diff
@@ -77,6 +77,40 @@ fmt = "black ."
 lint = "ruff ."
 test = "pytest . --cov anywidget --cov-report term-missing"
 
+# https://github.com/charliermarsh/ruff
+[tool.ruff]
+line-length = 88
+target-version = "py37"
+src = ["anywidget", "tests"]
+extend-select = [
+    "E",    # style errors
+    "F",    # flakes
+    "D",    # pydocstyle
+    "I",    # isort
+    "UP",   # pyupgrade
+    "S",    # bandit
+    "C4",   # flake8-comprehensions
+    "B",    # flake8-bugbear
+    "A001", # flake8-builtins
+    "TID",  # flake8-tidy-imports
+    "RUF",  # ruff-specific rules
+]
+extend-ignore = [
+    "D100", # Missing docstring in public module
+    "D104", # Missing docstring in public package
+    "D107", # Missing docstring in __init__
+    "D203", # 1 blank line required before class docstring
+    "D212", # Multi-line docstring summary should start at the first line
+    "D213", # Multi-line docstring summary should start at the second line
+    "D401", # Imperative mood
+    "D413", # Missing blank line after last section
+    "D416", # Section name should end with a colon
+]
```

Comment:
Just so I'm aware, should I read any of these as temporary ignores (i.e., eventually we should respect this rule) or not? These are all related to pydocstyle, correct?

Reply:
docstyle is a little frustrating, due to the multiple docstring conventions (numpy, google, etc...).  This config is for a numpy docs.  It should be noted that ruff now does provide this as

```toml
[tool.ruff.pydocstyle]
convention = "google"  # Accepts: "google", "numpy", or "pep257".
```

unfortunately, that _excludes_ D417 (`Missing argument description in the docstring: {name}`) which is to me the single most important docstring rule (it makes sure your docstrings actually match the parameters in the function).  I also think that there are a lot of annoying/invasive rules in the general D set (things like imperative mood or module docstrings can sometimes make you feel like you're spending more time appeasing the linter than doing useful things).  So this is the config I've been using in all my projects that seems to be the most useful / least invasive.  

That said, I totally agree... it's a lot.  Let me play with it a bit and see if I can get the important bits without the draconian bits in a minimal config

Reply:
I see., thanks for the explanation. Too bad there isn't an `extends` for a convention lol. But I agree D417 sounds like the most useful docstring rule, so let's keep it this way.

Don't worry about poking around with it too much, I trust the balance and having the comments is really useful. Maybe we could just add a comment to this like "numpy + D417" so I remember

Reply:
regarding extends for convention, i suspect it will come eventually: https://github.com/charliermarsh/ruff/issues/2606